crates/ra_cfg/src/cfg_expr.rs

```diff
@@ -33,6 +33,36 @@ impl CfgExpr {
             CfgExpr::Not(pred) => pred.fold(query).map(|s| !s),
         }
     }
+
+    /// Return minimal features needed
+    pub fn minimal_features_needed(&self) -> Option<Vec<SmolStr>> {
+        let mut features = vec![];
+        self.collect_minimal_features_needed(&mut features);
+        if features.is_empty() {
+            None
+        } else {
+            Some(features)
+        }
```

Comment:
I don't why the vector must not be empty, but if this invariant does need to be hold it's better to encode this contract with a newtype:
```rust
struct Features(Vec<SmolStr>);
impl Features {
    fn new(feature_names: Vec<SmolStr>) -> Option<Self> {
        match feature_names.is_empty() {
            true => None,
            false => Self(feature_names)
        }
    }
}
```

Reply:
I choose to return just a vector, you're right option here could be unnecessary.